src/tests/catwalk_tests/test_evaluation.py

```diff
@@ -1,22 +1,243 @@
+from tests.utils import fake_trained_model, MockMatrixStore
+from tests.results_tests.factories import (
+    ModelFactory,
+    TestPredictionFactory,
+    TrainPredictionFactory,
+    SubsetFactory,
+    init_engine,
+    session,
+)
 
 
 @Metric(greater_is_better=True)
 def always_half(predictions_proba, predictions_binary, labels, parameters):
     return 0.5
 
 
+SUBSETS = [
+    {
+        "name": "evens",
+        "query": """\
+            select entity_id, as_of_date
+            from {results_schema}.predictions
+            where entity_id % 2 = 0
+            and as_of_date in (SELECT(UNNEST(ARRAY{as_of_dates}::timestamp[])))
+            and model_id = '{model_id}'
+        """,
+        "hash": "klmno",
+    },
+    {
+        "name": "odds",
+        "query": """\
+            select entity_id, as_of_date
+            from {results_schema}.predictions
+            where entity_id % 2 = 1
+            and as_of_date in (SELECT(UNNEST(ARRAY{as_of_dates}::timestamp[])))
+            and model_id = {model_id}
+        """,
+        "hash": "fghij",
+    },
+    {
+        "name": "empty",
+        "query": """\
+            select entity_id, as_of_date
+            from {results_schema}.predictions
+            where entity_id = -1
+            and as_of_date in (SELECT(UNNEST(ARRAY{as_of_dates}::timestamp[])))
+            and model_id = {model_id}
+        """,
+        "hash": "abcde",
+    },
+]
+
+
+def setup_results_schema(db_engine, num_entities, labels):
+    # set up data, randomly generated by the factories but conforming
+    # generally to what we expect model_metadata and results schemas data
+    # to look like
+    model = ModelFactory()
+
+    class ImmediateTestPredictionFactory(TestPredictionFactory):
+        as_of_date = factory.LazyAttribute(
+            lambda o: o.model_rel.train_end_time
+        )
+    class ImmediateTrainPredictionFactory(TrainPredictionFactory):
+        as_of_date = factory.LazyAttribute(
+            lambda o: o.model_rel.train_end_time
+        )
+    
+    for entity in range(0, num_entities):
+        ImmediateTestPredictionFactory(
+            model_rel=model,
+            entity_id=entity,
+            label_value=labels[entity],
+        )
+        ImmediateTrainPredictionFactory(
+            model_rel=model,
+            entity_id=entity,
+            label_value=labels[entity],
+        )
+
+    session.commit()
+        
+    scores = []
+    labels = []
+    for row in db_engine.execute(
+        """\
+        select label_value, score
+        from test_results.predictions
+        order by entity_id
+        """
+    ):
+        labels.append(row[0])
+        scores.append(float(row[1]))
+
+    model_metadata = [
+        row 
+        for row in db_engine.execute(
+            """\
+            select model_id, train_matrix_uuid, train_end_time
+            from model_metadata.models
+            """
+        )
+    ][0]
```

Comment:
perhaps no clearer, but alternatively, (if clearer) --

```
(model_metadata,) = db_engine.execute(...)
```

-- assuming we expect one and only one row; or, if there really are more but you just want the first, *e.g.* --

```
model_metadata = db_engine.execute(...).first()
```

(in short, no need to collect all the results into a list when all you really want is the first result.)

Reply:
cool! didn't know about `.first()` that is much nicer!

Reply:
Function no longer exists in latest push.